xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp

```diff
@@ -1911,7 +1913,7 @@ bool CActiveAE::RunStages()
       float buftime = (float)(*it)->m_inputBuffers->m_format.m_frames / (*it)->m_inputBuffers->m_format.m_sampleRate;
       if ((*it)->m_inputBuffers->m_format.m_dataFormat == AE_FMT_RAW)
         buftime = (*it)->m_inputBuffers->m_format.m_streamInfo.GetDuration() / 1000;
```

Comment:
I am still wondering if the wrong logic is just happening here, cause half of the time should be counted at this stage. That way the twice the amoung of buffers would make it without waterlevel being reached.

Reply:
Good, except that not explains why jumps happens also in RAW that not uses half duration. Also in v19.0 IEC that not uses 12 audio units, etc. 

Although I tested this to confirm jumps stills happening and `GetDuration` in PAPlayer still needs divided by 2 (this last really was not expecting it).

Reply:
PaPlayer by itself uses a buggy concept ... the entire logic in there is flawed .. calculating one thing with samplerate and the other one with duration ... this can't work ... at least not for "half packages".

I hope to find some time over the weekend to restore one of my PCs that can do TrueHD again ... to find out, where in the stages we calculate it wrong ... PaPlayer I see a different issue. Fixing the "WATERLEVEL" might be possible - without fixing paplayer. In above equation I basically assume, that we for IEC:
- on this stage only should add up 10 ms
- that way the WATERLEVEL and time blocking will hold twice the number of packages
- Dividing by two for WATERLEVEL alone should not be needed anymore as here simply the time that was added up for comparing is now fixed.
- The real issue I think is that we only create "IEC frames" on the very last mile in the output of the sink. All the stream info and packages use / take the 10 ms version of things.

Reply:
> The real issue I think is that we only create "IEC frames" on the very last mile in the output of the sink. All the stream info and packages use / take the 10 ms version of things.

Then why happens also with RAW that uses 24 units / 20 ms in all stages?
Then why happens also in v18.5, v18.9, v19.0, v19.2 before MAT packer rework?

This PR fixes jumps and bad a/v sync in TrueHD in general. **(A)**

You talk about fix a different issue introduced later, only in IEC and that seems only affects PAPlayer calculations. **(B)**

Even in case to fix **B** we still needs fix **A**

Reply:
I think we are more on the same page as you think.

I think you agree with me that calculating a waterlevel based on 20 ms is totally wrong, when the packages we add at this stage are only 10 ms in length. Right? This we can properly fix and make sure that we in general only calculate with 10 ms TrueHD packages. As the only place where 20 ms come into the place is when a readily packed IEC packet is given into the Sink.

The way back we need to take care, as this obviously has then a 20 ms delay which AE is not allowed to compare to its internal "10 ms" packages ...

^^ The above written is a nobrainer and I think you directly agree with me. This we have to fix to get Delay computation correct over all the stages. 

I even want to say: As long as the above describe is not fixed by us ... every "a/v" sync will just be a hack somewhere. AE needs to know what is happening, else it cannot judge. The WATERLEVEL, fluctuating or not is one symptome for that.

For PaPlayer ... yeah - as you said no matter if entirely 20 ms or 10 / 20 ms mix it has an issue in not being smooth.

Don't feel blocked by my comments. I hopefully have time over the weekend to check the entire chain.

Reply:
AE measures ... times. It does not only trust package sizes it has an internal clock to know "how long stuff needed to be put on the sink" and compares a delay from that. There is definitely logical gap between what we currently "tunnel and fake" and what AE measures. It will always know that it needed 20 ms for a 10 ms packet (as it thinks it is 10 ms) at the last stage ... for this first issue AE has no "packet assumption", e.g. this 10 ms packet has a 10 ms overhead or something - it's not existing. It assumes a 1:1 relationship between packet the time it needs to write to the sink and the payload it writes.

Just to emphasize again on the first problem. PaPlayer for me is a different beast, cause as you said: It's there forever.

Emphasizing on the difference between DTS-HD-MA and TrueHD is - that we don't wait 61440 frames to create 24 packets at once. Here we work on a frame level. The old code _before_ when RAW came in, also worked on 1/24 units of TrueHD and therefore neither a 10ms implicit delay, nor a waiting for 61140 frames was needed.

So to answer your question: PaPlayer most likely was broken the very moment this 61440 workaround for the "RAW" TrueHD was introduced.

We can also solve it by dropping RAW support for TrueHD :P and work on a 1/24 unit level again :-)

Reply:
> I think you agree with me that calculating a waterlevel based on 20 ms is totally wrong, when the packages we add at this stage are only 10 ms in length. Right?

In AudioTrack RAW is 20 ms in all stages and waterlevel still wrong. Why?

Reply:
And btw. 61440/192000 == 320 ms ... this needs to be incorporated as an offset somewhere, cause this is an additional delay that we produce by waiting to have ful 24 units.

Reply:
All I want to say is: Here are multiple issues since many years and as long that is not fully understood - hacks for the waterlevel, hacks for the paplayer are needed ... and when those hacks work, users will still say: TrueHD is for me 250 ms out of sync ;-).

TrueHD fluctates, yes it does - as packages are "combined" and added only in bigger units at once - it has to fluctuate as the wallclock is ticking.
Why is the waterlevel increasing "factors" about 200 ms? Wrong calculations ... if there are 10 ms packets going in that stage - 20 frames are allowed. If those are 20 ms, 10 are allowed. If one adds 10 and the other does: 10 * GetDuration(); then the time that is compared is obviously wrong, this was what I pointed out above. This is clearly a bug - one of them.

Reply:
On the one hand AE works on Duration Times (as there is nothing else you can do for RAW packages), but on the other it also works on a relation between sendframes (number of channels by 16 bit) / divided by samplerate.

PaPlayer is a nice example where one is compared / mixed with the other ... and this goes wrong for IEC (as of the new thing of 10 vs. 20) but also for RAW - cause size vs. the samplerate is not IEC padded / packed anymore and it's just wrong.

For RAW one needs to work time based. For IEC one needs to work time based AND frame based as there is a 1:1 relation ship.

Long story short: I hope to find time over the weekend to organize a TrueHD capable device to debug this.

Reply:
> And btw. 61440/192000 == 320 ms ... this needs to be incorporated as an offset somewhere, cause this is an additional delay that we produce by waiting to have ful 24 units.

What?   this is nonsense

61440 are bytes and TrueHD has 8 channels

61440 / 8 channels = 7860 / 2 (2 bytes or 16 bits every sample) = 3840 samples

3840 samples / 192000 samples/sec = 0.02 s = **20 ms**


**Method 2**

See _Basic Definitions_ in https://developer.dolby.com/globalassets/technology/dolby-truehd/dolbytruehdhighlevelbitstreamdescription.pdf

> The audio stream to be encoded is divided into a series of audio frames. Each audio frame contains
> the sampled data for one presentation unit. The duration of a presentation unit is **1/1200** second for
> audio sampled at multiples of 48 kHz and 1/1102.5 second for audio sampled at multiples of 44.1
> kHz

**1/1200** * 24 audio units = 0.02 = **20 ms**

Reply:
Jup :-) Missed the framesize above - sorry for that :-)

Reply:
Maybe you can help me: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp#L200
When I see this right - we only pack 12 Units into one 61440 frame. So that means this is the position where we actually carry only 10 ms payload but have it in a 20 ms frame?

Is this right?

Reply:
Then this is the position where suddenly our "frametime" goes factor 2. Before it was 24 Units inside one 61440 byte frame, but now it's only 10. Correct? Or in other words: We generate "two calls" for OutputPacket and compute therefore 2*20ms == 40 ms while only 2 * 10 ms are consumed to do so.

Reply:
We run full and should for TrueHD only return after two OutputPackage calls.

Reply:
> When I see this right - we only pack 12 Units into one 61440 frame. So that means this is the position where we actually carry only 10 ms payload but have it in a 20 ms frame?
> 
> Is this right?

Yes and No....   61440 only is the max capacity of buffer, audio data is padded here and can be much lower:

these  are the sizes of each audio unit:

500 | 1200 | 100 | 400 | ...... until 12 audio units

`m_dataSize`  is used only as "flag" to signal packetComplete

if m_dataSize == 0   packet is not complete

if m_dataSize == TRUEHD_BUF_SIZE  packet is "complete"    complete menas here 12 audio units

true length of audio units is calculated here:

https://github.com/xbmc/xbmc/blob/7d010efbb843fbf34018e9a5761dd74a472a338d/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp#L953-L954

As each audio unit has a header that indicates his length (coded inside audio unit as raw data). 

As you know "standard" size of each audio unit is 2560 bytes  (61440 / 24) BUT some streams can use larger audio units sporadically to make possible high bitrate peaks with condition sum of lengths of all audio units in a MAT frame ==  61440. ie is possible an audio unit of 7000 bytes followed of more small audio units (< 2560 bytes).

This true `len` is used only in MAT packer:
`m_packer->Pack(m_sinkFormat.m_streamInfo, buffer[0] + of, len);`

Padding bytes are always distributed equally between each audio unit (not at the end). This is the reason why it is a very complex system: the length of the audio units is variable and the length of the padding is also variable and dependent on all the lengths of the 24 audio units. Assuming that a very large audio unit (7000 bytes) falls right at the end of the MAT frame, the part that does not fit is saved for the next MAT frame, so there is no guarantee that data at output of  MAT packer still aligned in groups of 24 with a correspondence 1:1 as input.

But everything that I have explained here is useless to solve the problem of the jumps in PAPlayer and Bad a/v sync because as I have already mentioned about 20 times in this PR it does not depend on whether we use 12 or 24 audio units or IEC or RAW. 😊

This (jumps) already happened in **v18.5**

Reply:
In fact, the video that I have bothered to record that shows the jumps is using **24 audio units**.

Reply:
I am not that much interested in the jumps of Paplayer. As I already mentioned what it computes is crap anyways ... by mixing bytes, samplerate and so on. You rightfully say: sucks, no matter if 24 units or 12 units.

I am concerned here:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp#L244

in combination with: 
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp#L204

From my understanding, sorry again. This will produce "stuff" that needs 20 ms to get down the sink, the sink does not care a single bit about the content in such a frame ... it only cares how long does it need to process that data in the given configuration (8 channel 16 bit 192 khz) ... as (61440/16/192000) == 20 ms. This you compensate in this PR by adding 10 ms delay per package on top ... fine.

But the way back (!) does not fit. Somewhere someone needs to know, that "GetDuration()" of 20 ms does not mean that 20 ms of data that we currently had is now gone, but only 10 ms of them, while the rest being padding. So - the sink will always answer delay for 20 ms ... while the payload we currently have and fill is only 10 ms.

So let's go ahead and fix the paplayer scenario ... as said above hopefully weekend allows me to get that stuff reproduced ... to step through all stages.

Reply:
> I am not that much interested in the jumps of Paplayer.

Sorry but this is PR fixes this and nothing more. 
Maybe you shouldn't be reviewing the PR if you're not interested in the subject and don't have/use any TrueHD equipment. 

> sink will always answer delay for 20 ms ... while the payload we currently have and fill is only 10 ms.

Sink only handles MAT (IEC) and always has 24 units even when at INPUT of MAT packer has 12 units

----> 12 units + 12 units **input** ------> MAT packer --------> **output** 24 units ---------> Sink

Reply:
> This PR fixes jumps and bad a/v sync in TrueHD in general. (A)
> 
> You talk about fix a different issue introduced later, only in IEC and that seems only affects PAPlayer calculations. (B)
> 
> Even in case to fix B we still needs fix A
> 
> The good news is that in the case of fixing B we can drop the last commit (PAPlayer)

Reply:
This PR fixes **A**..... if you want can create another PR to fix **B** and drop last commit in this PR

Reply:
oO -> go ahead ... if something is "fixed" - one cannot fix it again ...

Reply:
![image](https://user-images.githubusercontent.com/1089309/218250947-64e0264a-64ce-442c-aef5-45962c9c2c44.png)

Finally found some time to check at the WaterLevel on a TrueHD setup. Yeah ... it is as expected most of the time exactly twice the 200 ms ... cause of AE adding 20 x 10 ms packages but GetDuration tells 20 ms ...

Reply:
Good debugged further. Has absolutely nothing to do with 10 ms or 20 ms packages - it just converges to the MAX_CACHE_TIME which is 400 ms  -> all good.

Key-Question I have is the following: We basically make AE ignore WL for TrueHD. AE assumes that when the CACHE is fully filled, or at least filled more than MAX_WATER_LEVEL no correction is needed. But as TrueHD "fluctuates" quite a lot in time, we want that AE looks for TrueHD more often.

Therefore my question: Do we want to make something like this: https://github.com/fritsch/xbmc/commit/c7c6720f9381d973c951da20c611cc2941819a64

Reply:
Reason: I want to avoid that short term effects will cause and AdjustError for everyone.

Reply:
Same here:

**12 audio units**

```
2023-02-11 11:23:38.347 T:12892    info <general>: ActiveAE::GetWaterLevel 0.280000
2023-02-11 11:23:38.355 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.355 T:12892    info <general>: ActiveAE::GetWaterLevel 0.300000
2023-02-11 11:23:38.355 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.355 T:12892    info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:23:38.355 T:12892    info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:23:38.356 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.356 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.356 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.356 T:12892    info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:23:38.356 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.356 T:12892    info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:23:38.397 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.397 T:12892    info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:23:38.399 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.399 T:12892    info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:23:38.399 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.399 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.400 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.400 T:12892    info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:23:38.400 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.400 T:12892    info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:23:38.447 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.447 T:12892    info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:23:38.448 T:12892    info <general>: ActiveAE::GetWaterLevel 0.280000
2023-02-11 11:23:38.454 T:12892    info <general>: Skipped 1 duplicate messages..

```

**24 audio units**

```
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.280000
2023-02-11 11:24:34.636 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.300000
2023-02-11 11:24:34.636 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:24:34.636 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:24:34.637 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.637 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.637 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.637 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.637 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.637 T:1964     info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:24:34.662 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.662 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.662 T:1964     info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:24:34.668 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.668 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.669 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.669 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.669 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.669 T:1964     info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:24:34.712 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.712 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.723 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.723 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.723 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.723 T:1964     info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:24:34.762 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
```

But MAIN point is that this behavior has nothing related to use 12 or 24 audio units.

Reply:
Yes, see above. You are fully correct. See my suggestion. Maybe it benefits us.

Reply:
> Do we want to make something like this:

Sorry this not works: still jumping

Reply:
Could you show me in between which values it is jumping? What is the standard deviation? Really 200 Ms at once? Here in also I don't see such heavy jumps, see also the image above

Reply:
![image](https://user-images.githubusercontent.com/1089309/218255854-05c86c2e-eeee-42de-b8d1-bf40b8c9ba69.png)
From my image - I see that there is a max delta of 150 between GetWaterLevel calls. So Adjust would with my idea happen only every 100 ms jump, while the old code would jump in when waterlevel below 200 ms.

What  really want to avoid is, that the P-Regulator would bounce with the input - that way you would micro correct, which causes the opposite of what you want to achieve.

Reply:
I don't want to micro discuss this issue anymore. If you would be so kind and change ignoreWL to "alwaysAdjust" with a little comment of what the issue is with TrueHD frames appearing I would already be happy :-) and we can continue to merge ... or better: test build for the Forum people.

Reply:
https://user-images.githubusercontent.com/58434170/218256050-25fc4604-2214-4841-8766-b1833bae40af.mp4

Reply:
Okay - that's aweful ... then let's continue as it is right now ... thanks much.

Reply:
> or better: test build for the Forum people.

Forum feedback at this point is almost useless because now everyone there has their advancedsettings set to 90 or 100 and another negative delay in advancedsettings or GUI settings to "compensate".

At best, people are aware that they are wearing it. At worst he won't even be aware. That is, from now everyone will have a/v sync delay issues even those who have never had them :)

Reply:
I think they will realize the improved A/V sync ... or realizing that now a shorter "compensate" also already helps. Or they will realize that AE now earlier adjusts and maybe earlier causes ErrorAdjust. So let's stay tuned. I posted the latest version in the forum.